support custom deflate, inflate compressed bodies

trxcllnt opened the pull request:
This PR adds support for providing custom implementations of `createGzip` and `createDeflate` and inflating already-compressed response bodies if the client doesn't support compression.

Overriding `createGzip` allows swapping out gzip with [`pigz`](https://github.com/madler/pigz), useful for multi-megabyte dynamic payloads:
```js
const es = require('event-stream')
const { spawn } = require('child_process')
const pigz = { createGzip: () => es.child(spawn(`pigz`, ['-c'])) }
fastify.register(compressPlugin, { global: false, zlib: pigz })
```

Inflating compressed content is a convenience feature for services interfacing with clients like `curl`, or embedded devices, especially in a mesh scenario where data is produced and consumed deflated until it hits the edge. That said, I recognize it adds an extra stream between `reply.send` and the network interface, so I'd be down for making it opt-in.

mcollina commented on index.js:
```diff
@@ -5,6 +5,9 @@ const zlib = require('zlib')
 const mimedb = require('mime-db')
+const isStream = require('is-stream')
+const intoStream = require('into-stream')
+const maybeUnzip = require('unzip-if-zip')
```
I'm not very comfortable in adding a dependency on this module. Are there any alternatives?

trxcllnt replied:
yeah it seemed sus. I checked out the source, it's fairly simple to re-implement. probably worth doing just to add support for brotoli too.

mcollina commented on index.js:
```diff
@@ -81,13 +92,21 @@ function compressPlugin (fastify, opts, next) {
+      payload = payloadAsStream(payload)
+      stream = maybeUnzip()
+      pump(payload, stream, onEnd.bind(reply))
+      return next(null, stream)
```
I would prefer if we did _not_ convert it to a Stream if we do not have to unzip.

trxcllnt replied:
Not a problem. If we extract out the `unzip-if-zip` logic into a standalone function, we can call it if the payload is a string or buffer.

mcollina replied:
go for it!

mcollina commented on index.js:
```diff
@@ -81,13 +92,21 @@ function compressPlugin (fastify, opts, next) {
+      payload = payloadAsStream(payload)
+      stream = maybeUnzip()
+      pump(payload, stream, onEnd.bind(reply))
+      return next(null, stream)
```
same here.

trxcllnt commented:
@mcollina I updated the implementation to remove unzip-if-zip, avoid converting inflated payloads into streams, and placed the auto-inflating logic behind a feature flag because it's a breaking change for anybody sending deflated payloads today (like static file servers).

Also [it seems](https://stackoverflow.com/a/39032023/3117331) there's no reliable way to detect whether a stream is brotli encoded without attempting to decode the entire payload, otherwise I would've liked to add that too.

mcollina commented on test.js:
```diff
@@ -178,6 +288,71 @@ test('should not compress on missing header', t => {
+//     t.notOk(res.headers['content-encoding'])
+//     t.deepEqual(JSON.parse('' + res.payload), json)
+//   })
+// })
```
why is this commented out?

trxcllnt replied:
Mostly to draw attention, see the comment above:
>
```
// This throws an error, because the Reply's `send` method attempts
// to JSON.parse the compressed string before the `onSend` hook is invoked
```
>
Looking back I think I meant to say JSON.stringify instead of JSON.parse, but either way this test fails because `reply.send()` attempts to serialize the stringified gzipped buffer. It seems pretty edge-casey that anyone would do this, but I added the test for completeness.

mcollina replied:
I'm a bit lost. This test should pass because a buffer is sent as-is: https://github.com/fastify/fastify/blob/4b5f818b08d0eea02c46c782281fe90ff596830a/lib/reply.js#L65-L70.

trxcllnt replied:
In this test the payload is the utf8 string of the gzipped Buffer:
>
```
zlib.gzipSync(JSON.stringify(json)).toString()
```
>

mcollina replied:
Strings are trated the same. Why is this failing?

trxcllnt replied:
I dug deeper, and it looked like utf8-encoding the gzipped buffer and decoding only the first 4 characters was producing a Buffer that didn't pass the `isGzip()` check, and falling through to the normal `reply.serialize()` logic. So I removed the auto-inflate of strings when there's no header, since I doubt we want to sniff the encoding and decode every string that gets sent out.

trxcllnt commented on index.js:
```diff
@@ -5,6 +5,15 @@ const zlib = require('zlib')
+const isStream = require('is-stream')
+const intoStream = require('into-stream')
+const peek = require('peek-stream')
+const through = require('through2')
```
Sure can. The original unzip-if-zip module used through2, so I didn't change it. What about using [`minipass`](https://www.npmjs.com/package/minipass) instead of `PassThrough`, since this is a hot path? I think it should work, unless you can't pipe a readable that doesn't implement `unpipe()` to an http.ServerResponse.

mcollina replied:
go for it!

trxcllnt commented:
@mcollina sorry it's taken me so long to revisit this. I rebased this branch from master, updated to use minipass, and removed the weird edge case around string inflating. Let me know if there's anything else I should add, thanks.

cemremengu commented on package.json:
```diff
@@ -6,8 +6,17 @@
+    "pump": "^3.0.0",
+    "pumpify": "^1.3.3",
+    "string-to-stream": "^1.1.0",
+    "unzipper": "^0.8.9"
```
Too many additional dependencies but I guess thats just the way it is in ecosystem :smile:

trxcllnt commented on index.js:
```diff
@@ -156,6 +181,58 @@ function shouldCompress (type, compressibleTypes) {
+function unzipStream (inflate, maxRecursion) {
+  if (!(maxRecursion >= 0)) maxRecursion = 3
+  return peek({ newline: false, maxBuffer: 10 }, function (data, swap) {
+    if (maxRecursion < 0) return swap(new Error('Maximum recursion reached'))
```
This code was from the `unzip-if-zip` module, I believe it's just guarding against the input has been run through deflate > 3 times, or somehow the inflated payload's first `n` bytes still include the deflate prefix. I can try to add a test for that if you want.

cemremengu replied:
@mcollina wdyt?

mcollina replied:
Yes, adding a test case would be good.
